app/vmagent/remotewrite/client.go

```diff
@@ -462,21 +462,27 @@ again:
 		// - Real-world implementations of v1 use both 400 and 415 status codes.
 		// See more in research: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/8462#issuecomment-2786918054
 	} else if statusCode == 415 || statusCode == 400 {
-		if c.canDowngradeVMProto.Swap(false) {
-			logger.Infof("received unsupported media type or bad request from remote storage at %q. Downgrading protocol from VictoriaMetrics to Prometheus remote write for all future requests. "+
-				"See https://docs.victoriametrics.com/victoriametrics/vmagent/#victoriametrics-remote-write-protocol", c.sanitizedURL)
-			c.useVMProto.Store(false)
-		}
-
 		if encoding.IsZstd(block) {
 			logger.Infof("received unsupported media type or bad request from remote storage at %q. Re-packing the block to Prometheus remote write and retrying."+
 				"See https://docs.victoriametrics.com/victoriametrics/vmagent/#victoriametrics-remote-write-protocol", c.sanitizedURL)
 
-			block = mustRepackBlockFromZstdToSnappy(block)
+			zstdBlock := block
+			block, err = repackBlockFromZstdToSnappy(zstdBlock)
+			if err == nil {
+				if c.canDowngradeVMProto.Swap(false) {
+					logger.Infof("received unsupported media type or bad request from remote storage at %q. Downgrading protocol from VictoriaMetrics to Prometheus remote write for all future requests. "+
+						"See https://docs.victoriametrics.com/victoriametrics/vmagent/#victoriametrics-remote-write-protocol", c.sanitizedURL)
+					c.useVMProto.Store(false)
+				}
+
+				c.retriesCount.Inc()
+				_ = resp.Body.Close()
+				goto again
+			}
 
-			c.retriesCount.Inc()
-			_ = resp.Body.Close()
-			goto again
+			logger.Warnf("failed to repack zstd block (%s) to snappy; The block will be rejected. "+
+				"Possible cause: ungraceful shutdown leading to persisted queue corruption.",
+				len(zstdBlock), err)
```

Comment:
Wdyt about also incrementing metric for data dropped items?
That would allow to set up alerts for such events if needed.

Reply:
It is incremented a few lines below.